### PR TITLE
feat: use updated pubsub time cache implementation

### LINKED ISSUE #9554: Flaky test: ci/gh-experiment: go test: pubsub_msg_seen_cache_test
Creating issue so others don't spend time investigating.

`TestMessageSeenCacheTTL` sometimes fails on GitHub version of our CI (log below).

@galargh  I suggest we wait with investigation until https://github.com/ipfs/kubo/pull/9543 lands, as these tests will be refactored by the mentioned PR anyway.

https://github.com/ipfs/kubo/actions/runs/3926432440/jobs/6712202879#step:5:11887:
```
=== FAIL: test/integration TestMessageSeenCacheTTL (1.54s)

Computing default go-libp2p Resource Manager limits based on:
    - 'Swarm.ResourceMgr.MaxMemory': "1.8 GB"
    - 'Swarm.ResourceMgr.MaxFileDescriptors': 32768

Applying any user-supplied overrides on top.
Run 'ipfs swarm limit all' to see the resulting limits.


Computing default go-libp2p Resource Manager limits based on:
    - 'Swarm.ResourceMgr.MaxMemory': "1.8 GB"
    - 'Swarm.ResourceMgr.MaxFileDescriptors': 32768

Applying any user-supplied overrides on top.
Run 'ipfs swarm limit all' to see the resulting limits.


Computing default go-libp2p Resource Manager limits based on:
    - 'Swarm.ResourceMgr.MaxMemory': "1.8 GB"
    - 'Swarm.ResourceMgr.MaxFileDescriptors': 32768

Applying any user-supplied overrides on top.
Run 'ipfs swarm limit all' to see the resulting limits.

    pubsub_msg_seen_cache_test.go:136: bootstrap peer=QmeDXb79V25BrSP4ENDvqZMd7qBr7SXuKSNwxxQawjdbwK, consumer peer=QmaxMa42XFxgG8Rn6TPCvsSKfZ4dbxxwnstt4AK8xaqUqS, producer peer=QmZWxchjiixjTNgh5TA8XZFrYfQPSaDkiRRH8AFCkmy41e
    pubsub_msg_seen_cache_test.go:114: sending [msg_1] with duplicate message ID at [Jan 16 02:10:31.625]
    pubsub_msg_seen_cache_test.go:191: did not receive [msg_1] by [Jan 16 02:10:32.626]
    pubsub_msg_seen_cache_test.go:192: context deadline exceeded
```

### PR DESCRIPTION
**Pull request by smrz2001**
This PR closes https://github.com/ipfs/kubo/issues/9542 – attempts to apply the fix for https://github.com/libp2p/go-libp2p-pubsub/issues/502 to `kubo`.

Once an official pubsub release is available, this draft PR can be updated accordingly and marked ready for review.

cc @lidel @BigLep
```
sending [msg_1] with message ID [MsgId_1] at T0.000000s
received [msg_1] at T0.000726s
sending [msg_2] with message ID [MsgId_1] at T0.002784s
did not receive [msg_2] at T1.011944s
sending [msg_3] with message ID [MsgId_2] at T1.013189s
received [msg_3] at T1.014176s
sending [msg_4] with message ID [MsgId_1] at T9.913194s
did not receive [msg_4] at T10.924792s
sending [msg_5] with message ID [MsgId_1] at T10.930913s
did not receive [msg_5] at T11.944472s
sending [msg_6] with message ID [MsgId_2] at T11.950365s
received [msg_6] at T11.950365s
sending [msg_7] with message ID [MsgId_1] at T21.036650s
received [msg_7] at T21.037893s
sending [msg_8] with message ID [MsgId_3] at T21.039733s
received [msg_8] at T21.039733s
--- PASS: TestMessageSeenCacheTTL (21.66s)
```

Should also close #9554

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -1,5 +1,7 @@
 module github.com/ipfs/kubo
 
+replace github.com/libp2p/go-libp2p-pubsub => github.com/smrz2001/go-libp2p-pubsub v0.0.0-20230110155724-04bfcf58514f
```
Comment (lidel) [inline go.mod:3]: FYSA I've pushed this change, so we can see if CI passes.
(It is late Friday so will do proper review on Monday)

@smrz2001 In the future, you can point at an arbitrary commit from your own fork via:

```console
$  go mod edit -replace github.com/libp2p/go-libp2p-pubsub=github.com/smrz2001/go-libp2p-pubsub@04bfcf58514f59bddb650fec5b2edac018c8c406
```

Comment (smrz2001) [reply go.mod:3]: Thanks, so much, @lidel!! I didn't know we could point to fork commits like that!

Comment (smrz2001) [reply go.mod:3]: I'll fix any CI issues that come up.

**Comment by lidel** [comment]
Updated to include our GH version of CI from `master` and see if this PR fixes issue described in https://github.com/ipfs/kubo/issues/9554.

Will check on the result tomorrow.

**Comment by smrz2001** [comment]
@lidel, since @vyzo requested switching back to the original, "first seen" time cache implementation in the pubsub [PR](https://github.com/libp2p/go-libp2p-pubsub/pull/513), I updated the `kubo` code to explicitly use the new, "last seen" implementation.

I could also make this configurable if you suggest. Regardless, I think it would be best to default to the new implementation for `kubo`, even if configurable.

I could also make it configurable in a follow up PR. Please let me know either way.

**Comment by smrz2001** [comment]
@lidel, I see the "context deadline timeout" failure on (my previous) flaky pubsub tests that you mentioned might get fixed via this PR.

I'm going to try bumping the timeout to see if that helps. 1 second is probably too small.

**Comment by smrz2001** [comment]
@lidel, running `go get github.com/libp2p/go-libp2p-pubsub/timecache` as recommended by the CI failures causes `go-graphsync` to get downgraded and a bunch of other things to fail.

I'm going to wait till @vyzo makes the official pubsub release to point to that instead of my fork, and then run the `go get` command.

**Comment by smrz2001** [comment]
> @lidel, I see the "context deadline timeout" failure on (my previous) flaky pubsub tests that you mentioned might get fixed via this PR.
> 
> I'm going to try bumping the timeout to see if that helps. 1 second is probably too small.

Bah, it failed even with the larger value (though it passed just before that with the same value) :(

**Review by lidel — CHANGES_REQUESTED** [review]
> I could also make this configurable if you suggest. Regardless, I think it would be best to default to the new implementation for kubo, even if configurable.
>  I could also make it configurable in a follow up PR. Please let me know either way.

Fine to change the default, but please include a configuration option in this PR – @vyzo makes a good point, there may be people who build things based on the old behavior

We should document the change in release notes, allowing them to switch back via optional configration option.

Let's make it future-proof, in case more advanced strategies will be added in the future.
Add `*OptionalString` to [config/pubsub.go](https://github.com/ipfs/kubo/blob/master/config/pubsub.go),  and `docs/config.md`, something like:

```diff
+SeenExpirationStrategy *OptionalString  `json:",omitempty"`
```

With `last-observed` being the new implicit default and `first-observed`  being the old one (feel free to pick better names). :pray:

**Comment by smrz2001** [comment]
@lidel, I just pushed some updates to make the cache implementation configurable. I relied on the enum I added to the pubsub code and used an `OptionalInteger` for the new config (which for now would be 0 (first seen) or 1 (last seen)). Please let me know if I need to change this. I can just add strings that map to the enum values also.

**Comment by smrz2001** [comment]
> @lidel, I just pushed some updates to make the cache implementation configurable. I relied on the enum I added to the pubsub code and used an `OptionalInteger` for the new config (which for now would be 0 (first seen) or 1 (last seen)). Please let me know if I need to change this. I can just add strings that map to the enum values also.

Ok, nvm, I changed it to a string.

**Comment by BigLep** [comment]
@smrz2001 : FYI that @lidel is going to prioritize the review here so that we can wrap this up today/tomorrow and do 0.18.1 tomorrow: https://github.com/ipfs/kubo/issues/9579

@lidel : I assume we'll do a changelog addition.  0.18.1 changelog has been started in https://github.com/ipfs/kubo/pull/9587

**Comment by smrz2001** [comment]
Thanks so much for the cleanup and tweaks, @lidel 🙏🏼

**Review by lidel — APPROVED** [review]
Thank you @smrz2001 for driving this and being patient with us.
Added some docs and release notes, merging.

0.18.1 release can be tracked in https://github.com/ipfs/kubo/issues/9579 (ETA Friday or Monday, depends if other item lands on time)

**Pull request merged**